File: baekjoon/그래프/[1260]DFS와BFS.py
from collections import deque

def dfs(graph, v, visited): # 그래프, 시작 번호, 방문 유무
    visited[v] = True
    print(v, end=" ")
    
    for i in graph[v]:
        if not visited[i]:
            dfs(graph, i, visited)
            
def bfs(graph, v, visited):
    queue = deque([v])
    visited[v] = True
    
    while queue:
        a = queue.popleft()
        print(a, end=" ")
        
        for i in graph[a]:
            if not visited[i]:
                queue.append(i)
                visited[i] = True            
    
n, m, v = map(int, input().split()) # 정점, 간선, 시작 번호
graph = [[] for i in range(n+1)]

for i in range(m):
    start, end = map(int, input().split())
    graph[start].append(end)
    graph[end].append(start)
    
for i in range(1, len(graph)):
    graph[i] = sorted(graph[i])

visited = [False]*(n+1)
dfs(graph, v, visited)    
print()    
visited = [False]*(n+1)
bfs(graph, v, visited)

File: baekjoon/그래프/[1260]DFS와BFS.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
using namespace std;

void dfs(int start, vector<int> graph[], bool check[]) {
    check[start] = 1;
    cout << start << " ";

    for (int i = 0; i < graph[start].size(); i++) {
        int next = graph[start][i];

        if (!check[next]) {
            dfs(next, graph, check);
        }
    }
}

void bfs(int start, vector<int> graph[], bool check[]) {
    queue<int> q;   

    q.push(start);
    check[start] = 1;

    while (!q.empty()) {
        int temp;
        temp = q.front();
        q.pop();

        cout << temp << " ";

        for (int i = 0; i < graph[temp].size(); i++) {
            if (!check[graph[temp][i]]) {
                q.push(graph[temp][i]);
                check[graph[temp][i]] = 1;
            }
        }

    }

}

int main() {
    int n, m, start;
    cin >> n >> m >> start;
    vector<int> graph[1001];
    bool check[1001];

    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }

    for (int i = 1; i <= n; i++) {
        sort(graph[i].begin(), graph[i].end());
    }

    fill(check, check + n + 1, false);
    dfs(start, graph, check);
    
    cout << "\n";

    fill(check, check + n + 1, false);
    bfs(start, graph, check);
}